### PR TITLE
Extract ReaderPool from IndexWriter

### PR DESCRIPTION
**Pull request by s1monw**
ReaderPool plays a central role in the IndexWriter pooling NRT readers
and making sure we write buffered deletes and updates to disk. This class
used to be a non-static inner class accessing many aspects including locks
from the IndexWriter itself. This change moves the class outside of IW and
defines it's responsiblity in a clear way with respect to locks etc. Now
IndexWriter doesn't need to share ReaderPool anymore and reacts on writes done
inside the pool by checkpointing internally. This also removes acquiring the IW
lock inside the reader pool which makes reasoning about concurrency difficult.

This change also add javadocs and dedicated tests for the ReaderPool class.

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/index/IndexWriter.java [hunk 1]
```diff
@@ -536,338 +528,61 @@ public final long ramBytesUsed() {
...
 
-                // Only acquire IW lock on each write, since this is a time consuming operation.  This way
-                // other threads get a chance to run in between our writes.
-                synchronized (IndexWriter.this) {
-                  if (rld.writeFieldUpdates(directory, globalFieldNumberMap, bufferedUpdatesStream.getCompletedDelGen(), infoStream)) {
-                    checkpointNoSIS();
-                  }
+              // Only acquire IW lock on each write, since this is a time consuming operation.  This way
+              // other threads get a chance to run in between our writes.
+              synchronized (IndexWriter.this) {
```
Comment (dweiss) [inline lucene/core/src/java/org/apache/lucene/index/IndexWriter.java]: Could be just 'this' now, but I actually like the explicit form...

File: lucene/core/src/java/org/apache/lucene/index/IndexWriter.java [hunk 2]
```diff
@@ -4350,7 +4055,9 @@ synchronized private void _mergeInit(MergePolicy.OneMerge merge) throws IOExcept
     // Must move the pending doc values updates to disk now, else the newly merged segment will not see them:
     // TODO: we could fix merging to pull the merged DV iterator so we don't have to move these updates to disk first, i.e. just carry them
     // in memory:
-    readerPool.writeDocValuesUpdatesForMerge(merge.segments);
+    if(readerPool.writeDocValuesUpdatesForMerge(merge.segments)) {
```
Comment (dweiss) [inline lucene/core/src/java/org/apache/lucene/index/IndexWriter.java]: space after an if?

File: lucene/core/src/java/org/apache/lucene/index/IndexWriter.java [hunk 3]
```diff
@@ -5260,6 +4957,22 @@ public final int numDeletesToMerge(SegmentCommitInfo info) throws IOException {
...
+  }
+
+  void release(ReadersAndUpdates readersAndUpdates) throws IOException {
+    release(readersAndUpdates, true);
+  }
 
+  private void release(ReadersAndUpdates readersAndUpdates, boolean assertLiveInfo) throws IOException {
+    assert Thread.holdsLock(this);
+    if (readerPool.release(readersAndUpdates, assertLiveInfo)) {
+      assert Thread.holdsLock(this); // if we write anything here we have to hold the lock otherwise IDF will deletes files underneath us
```
Comment (dweiss) [inline lucene/core/src/java/org/apache/lucene/index/IndexWriter.java]: will deletes -> will delete

File: lucene/core/src/java/org/apache/lucene/index/ReaderPool.java [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+  private final Directory originalDirectory;
+  private final FieldInfos.FieldNumbers fieldNumbers;
+  private final LongSupplier completedDelGenSupplier;
+  private final InfoStream infoStream;
+  private final SegmentInfos segmentInfos;
+  private final String softDeletesField;
+  // This is a "write once" variable (like the organic dye
+  // on a DVD-R that may or may not be heated by a laser and
+  // then cooled to permanently record the event): it's
+  // false, until getReader() is called for the first time,
```
Comment (dweiss) [inline lucene/core/src/java/org/apache/lucene/index/ReaderPool.java]: setting of this flag has been moved to enableReaderPooling and it's used in tests, for example so maybe the doc should reflect this?

File: lucene/core/src/test/org/apache/lucene/index/TestReaderPool.java [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+    FieldInfos.FieldNumbers fieldNumbers = buildIndex(directory);
+    StandardDirectoryReader reader = (StandardDirectoryReader) DirectoryReader.open(directory);
+    SegmentInfos segmentInfos = reader.segmentInfos.clone();
+
+    ReaderPool pool = new ReaderPool(directory, directory, segmentInfos, fieldNumbers, () -> 0l, null, null, null);
+    SegmentCommitInfo commitInfo = RandomPicks.randomFrom(random(), segmentInfos.asList());
+    ReadersAndUpdates readersAndUpdates = pool.get(commitInfo, true);
+    assertSame(readersAndUpdates, pool.get(commitInfo, false));
+    assertTrue(pool.drop(commitInfo));
+    if (random().nextBoolean()) {
```
Comment (dweiss) [inline lucene/core/src/test/org/apache/lucene/index/TestReaderPool.java]: does it make any difference if we just do it all the time here (not randomly)?

Comment (s1monw) [reply lucene/core/src/test/org/apache/lucene/index/TestReaderPool.java]: there might be a bug that requires it to be called more than once. that is what I am trying to test here too.

**Pull request merged**